src/geode/geosciences_io/mesh/grdecl_input.cpp

```diff
@@ -0,0 +1,290 @@
+                       + 4 * nx_ * ny_ * grid_coordinates[2]],
+                top_left_pillar );
+            return points;
+        }
+
+        geode::NNSearch3D::ColocatedInfo create_all_points(
+            absl::Span< const Pillar > pillars,
+            absl::Span< const double > depths )
+        {
+            std::vector< geode::Point3D > points;
+            points.reserve( 8 * nx_ * ny_ * nz_ );
+            for( const auto k : geode::Range{ nz_ } )
+            {
+                for( const auto j : geode::Range{ ny_ } )
+                {
+                    for( const auto i : geode::Range{ nx_ } )
+                    {
+                        const std::array< geode::index_t, 3 > grid_coordinates{
+                            i, j, 2 * k
+                        };
+                        for( const auto point :
+                            cell_points( grid_coordinates, pillars, depths ) )
+                        {
+                            points.push_back( point );
+                        }
+                    }
+                }
+            }
+            const auto collocated_mapping =
+                geode::NNSearch3D{ points }.colocated_index_mapping(
+                    geode::global_epsilon );
+            for( const auto& point : collocated_mapping.unique_points )
+            {
+                builder_->create_point( point );
+            }
+            return collocated_mapping;
+        }
+
+        void create_grid( absl::Span< const Pillar > pillars,
+            absl::Span< const double > depths )
+        {
+            const auto collocated_mapping =
+                create_all_points( pillars, depths );
+            for( const auto cell_id : geode::Range{ nx_ * ny_ * nz_ } )
+            {
+                builder_->create_hexahedron(
+                    { collocated_mapping.colocated_mapping[0 + 8 * cell_id],
+                        collocated_mapping.colocated_mapping[1 + 8 * cell_id],
+                        collocated_mapping.colocated_mapping[2 + 8 * cell_id],
+                        collocated_mapping.colocated_mapping[3 + 8 * cell_id],
+                        collocated_mapping.colocated_mapping[4 + 8 * cell_id],
+                        collocated_mapping.colocated_mapping[5 + 8 * cell_id],
+                        collocated_mapping.colocated_mapping[6 + 8 * cell_id],
+                        collocated_mapping
+                            .colocated_mapping[7 + 8 * cell_id] } );
+            }
+            builder_->compute_polyhedron_adjacencies();
+        }
+
+        std::array< geode::index_t, 4 > cell_pillars_id(
+            const std::array< geode::index_t, 3 >& grid_coordinates ) const
+        {
+            const geode::index_t number_of_vertices_on_lines = nx_ + 1;
+            const geode::index_t bottom_left_pillar_id =
+                grid_coordinates[0]
+                + ( number_of_vertices_on_lines ) * ( grid_coordinates[1] + 1 );
```

Comment:
```suggestion
                + number_of_vertices_on_lines * ( grid_coordinates[1] + 1 );
```
simplify readability